File: QrLinkPdf.Wasm/Ocr.fs
/// Browser OCR engine, backed by a real `TesseractEngine` running in-process
/// under WebAssembly - the same `Tesseract.CrossPlatform` package the
/// desktop CLI uses (see ../Ocr.fs), via the new
/// `Tesseract.Native.browser-wasm` + `Tesseract.CrossPlatform.SkiaSharp`
/// preview packages (see QrLinkPdf.Wasm.fsproj's own comment - not yet
/// published to nuget.org). Replaces the previous JS-interop bridge to
/// tesseract-wasm (wwwroot/ocr.js, removed): PdfQrLinker's OcrEngine slot is
/// just `SKBitmap -> OcrWord list`, and TesseractEngine.Process is
/// synchronous, so there's no interop layer left to write here at all - the
/// whole scan/link pipeline stays synchronous the same way it already was.
module QrLinkPdf.Wasm.Ocr

open System
open System.IO
open System.Threading.Tasks
open Microsoft.JSInterop
open SkiaSharp
open Tesseract
open QrLinkPdf

let mutable private engine: TesseractEngine option = None

/// Loads the English trained-data model and constructs the engine - fetches
/// ~4 MB, so this is the one unavoidably async step; `Process` itself is
/// plain and synchronous once this is done. Safe to call more than once;
/// a no-op once `engine` is set.
let init (jsInProcess: IJSInProcessRuntime) : Task<unit> =
    task {
        if engine.IsNone then
            // Same asset-host convention index.html's loadBootResource hook
            // (and the old ocr.js) used - the page's own origin for local
            // dev, the Cloudflare asset host when deployed (see
            // qrLinkPdfGetAssetBase's own comment for the local-dev
            // fallback).
            let assetBase = jsInProcess.Invoke<string>("qrLinkPdfGetAssetBase")
            use http = new Net.Http.HttpClient(BaseAddress = Uri(assetBase))
            let! data = http.GetByteArrayAsync("tessdata/eng.traineddata")
            Directory.CreateDirectory("/tessdata") |> ignore
            do! File.WriteAllBytesAsync("/tessdata/eng.traineddata", data)
            engine <- Some(new TesseractEngine("/tessdata", "eng", EngineMode.Default))
    }

/// A ScanOptions.OcrEngine backed by the already-loaded engine from `init`.
/// Returns no words rather than throwing if `init` never succeeded (a
/// network hiccup, say) - same "fail this one scan's OCR quietly" contract
/// the old JS bridge had.
let create () : SKBitmap -> OcrWord list =
    fun bitmap ->
        match engine with
        | None -> []
        | Some engine -> TesseractWords.ofBitmap engine bitmap

File: QrLinkPdf.Wasm/wwwroot/ocr.js
// Bridges to tesseract-wasm's low-level, synchronous OCREngine (not the
// high-level OCRClient, which runs in a Worker and is Promise-based - the
// whole point here is a synchronous call PdfQrLinker.link's existing
// synchronous pipeline can use directly, via IJSInProcessRuntime).
//
// Loading the engine and its trained-data model is the one unavoidably
// async step (fetching the WASM module and ~4 MB of model data); recognize()
// itself is plain and synchronous once that's done.
window.qrLinkPdfOcr = {
  engine: null,

  async init() {
    if (this.engine) {
      return;
    }

    // Same asset-host convention index.html's loadBootResource hook uses -
    // empty means same origin (local dev), set means the Cloudflare asset
    // host (deployed).
    const base = window.qrLinkPdfAssetBase || './';

    const { createOCREngine } = await import(base + 'tesseract-wasm/lib.js');
    this.engine = await createOCREngine();

    const modelResponse = await fetch(base + 'tessdata/eng.traineddata');
    const modelBytes = await modelResponse.arrayBuffer();
    this.engine.loadModel(new Uint8Array(modelBytes));
  },

  // `bytes` is a flat RGBA8888 pixel buffer - the .NET side has already done
  // any colour-space conversion, since that's cheap and unambiguous there,
  // and OCREngine.loadImage just wants {width, height, data} shaped like
  // ImageData, not an actual ImageData instance.
  recognize(bytes, width, height) {
    // The engine is loaded eagerly at app startup (see Main.fs) and again,
    // as a no-op safety net, right before a scan that wants it - but if it
    // genuinely never loaded (a network hiccup, say), fail this one scan's
    // OCR quietly rather than throwing on a null engine.
    if (!this.engine) {
      return [];
    }

    const data = new Uint8ClampedArray(bytes);
    this.engine.loadImage({ width, height, data });

    const words = this.engine.getTextBoxes('word');
    this.engine.clearImage();

    return words.map((w) => ({
      left: w.rect.left,
      top: w.rect.top,
      right: w.rect.right,
      bottom: w.rect.bottom,
      text: w.text,
    }));
  },
};
